### PR TITLE
Common schedulers refactoring

### PR DESCRIPTION
**Pull request by andrey-churkin**

### CODE REVIEW

**Comment by andrey-churkin** [comment]
Jenkins please retry a build

**Comment by andrey-churkin** [comment]
Jenkins please retry a build

File: nncf/common/pruning/schedulers.py [hunk 1]
```diff
@@ -10,112 +10,146 @@
...
+            - controller: Pruning algorithm controller.
+            - initial_pruning: Pruning level which already has been
+                applied to the model. It is the level at which the schedule begins.
+            - target_pruning: Pruning level at which the schedule ends.
+            - num_warmup_epochs: Number of epochs for model pre-training before pruning.
+            - num_pruning_epochs: Number of epochs during which the pruning level
+                is increased from `initial_pruning` to `target_pruning`.
+            - freeze_epoch: Zero-based index of the epoch from which the pruning
+                mask will be frozen and will not be trained.
+            - current_pruning: Pruning level for the `current_epoch`.
```
Comment (alexsu52) [inline nncf/common/pruning/schedulers.py:48]:
```suggestion
        Initializes the internal state of the pruning scheduler specified by:
            - controller: Pruning algorithm controller.
            - params: dict of the scheduler parameters

        Scheduler parameters:
            - initial_pruning: Pruning level which already has been
                applied to the model. It is the level at which the schedule begins.
            - target_pruning: Pruning level at which the schedule ends.
            - num_warmup_epochs: Number of epochs for model pre-training before pruning.
            - num_pruning_epochs: Number of epochs during which the pruning level
                is increased from `initial_pruning` to `target_pruning`.
            - freeze_epoch: Zero-based index of the epoch from which the pruning
                mask will be frozen and will not be trained.
            - current_pruning: Pruning level for the `current_epoch`.
```

File: nncf/common/pruning/schedulers.py [hunk 2]
```diff
@@ -10,112 +10,146 @@
...
-    def _set_pruning_level(self):
-        self.algo.set_pruning_rate(self.current_pruning_level)
+        :return: Pruning level that should be applied to the model.
+        """
+        raise NotImplementedError(
+            'PruningScheduler implementation must override _calculate_pruning_level method.')
 
-        if self.current_epoch >= (self.pruning_steps + self.num_init_steps):
-            self.algo.freeze()
+    def epoch_step(self, next_epoch: Optional[int] = None) -> None:
```
Comment (alexsu52) [inline nncf/common/pruning/schedulers.py:86]: It looks like this class is a base class. Could you add docstring for epoch_step and step methods too.

File: nncf/common/pruning/schedulers.py [hunk 3]
```diff
@@ -128,13 +162,13 @@ class ExponentialWithBiasPruningScheduler(PruningScheduler):
...
 
-    def _calc_pruning_level(self):
-        curr_pruning = self.a * np.exp(-self.k * (self.current_epoch - self.num_init_steps - 1)) + self.b
-        max_pruning = self.pruning_target
+    def _calculate_pruning_level(self) -> float:
+        curr_pruning = self.a * np.exp(-self.k * (self.current_epoch - self.num_warmup_epochs - 1)) + self.b
+        max_pruning = self.target_pruning
         return max_pruning if curr_pruning >= max_pruning else curr_pruning
 
     @staticmethod
```
Comment (alexsu52) [inline nncf/common/pruning/schedulers.py:174]: Could you format docstring for  `def _init_exp(E_max, P_min, P_max, D=1 / 8)` method

File: nncf/common/pruning/schedulers.py [hunk 4]
```diff
@@ -10,112 +10,146 @@
...
 
-        # Pruning rates
-        self.initial_pruning = self.algo.pruning_init
-        if self.algo.prune_flops:
-            self.pruning_target = self._params.get('pruning_flops_target')
+        if self.controller.prune_flops:
+            self.target_pruning = params.get('pruning_flops_target')
         else:
-            self.pruning_target = self._params.get('pruning_target', 0.5)
+            self.target_pruning = params.get('pruning_target', 0.5)
```
Comment (vshampor) [inline nncf/common/pruning/schedulers.py:60]: Does not look well. Now the name of the parameter in the config and the name of the attribute differ. This is not critical, but affects debuggability.

File: nncf/common/pruning/schedulers.py [hunk 5]
```diff
@@ -10,112 +10,146 @@
...
-        super().epoch_step(next_epoch)
-        self._set_pruning_level()
+    def _maybe_freeze(self) -> None:
+        """
+        Checks if pruning mask needs to be frozen and not to be trained.
+        If freezing is needed, then the internal state of the `controller`
+        object will be changed.
+        """
+        if self.current_epoch >= self.freeze_epoch:
+            self.controller.freeze()
```
Comment (vshampor) [inline nncf/common/pruning/schedulers.py:74]: You saved a single line of code at a high cost of code readability. Suggest explicitly inlining these function calls.

File: beta/tests/tensorflow/sparsity/test_common.py [hunk 6]
```diff
@@ -176,19 +176,19 @@ def test_polynomial_schedule_per_epoch_step(self,
...
         self.run_epoch(steps_per_epoch, scheduler, mock, ref_sparsity_levels, explicit, epoch)
-        assert scheduler.current_sparsity_level == ref_sparsity_levels[3]
+        assert scheduler.sparsity_level == ref_sparsity_levels[3]
         epoch += 1
 
         # After epoch 3 - sparsity freeze should occur
         self.run_epoch(steps_per_epoch, scheduler, mock, ref_sparsity_levels, explicit, epoch)
         magnitude_algo_mock.freeze.assert_called_once()
-        assert scheduler.current_sparsity_level == ref_sparsity_levels[4]
+        assert scheduler.sparsity_level == ref_sparsity_levels[4]
```
Comment (vshampor) [inline beta/tests/tensorflow/sparsity/test_common.py:185]: With schedulers, we've got two major kinds of levels - the target level (of either sparsity or pruning) and the current level. When I see `sparsity_level` or `pruning_level`, it is not immediately obvious to me based on the naming which level is meant, while it *should* be immediately obvious since the mixup between the two types of levels may lead to subtle, hard to debug errors. You should keep both `current` and `target` portions in the level-related attribute name in all affected schedulers.

**Review by vshampor — CHANGES_REQUESTED** [review]

File: beta/tests/tensorflow/sparsity/test_common.py (hunk 6 above)
Comment (andrey-churkin) [reply beta/tests/tensorflow/sparsity/test_common.py]: I agree with you. I fixed it.

File: nncf/common/pruning/schedulers.py (hunk 5 above)
Comment (andrey-churkin) [reply nncf/common/pruning/schedulers.py]: I agree with you. I fixed it.

File: nncf/common/sparsity/schedulers.py [hunk 7]
```diff
@@ -11,207 +11,289 @@
...
         self.controller = controller
         self.initial_sparsity = self.controller.get_sparsity_init()
-        self.target_sparsity = self._params.get('sparsity_target', 0.5)
-        self.target_epoch = self._params.get('sparsity_target_epoch', 90)
-        self.freeze_epoch = self._params.get('sparsity_freeze_epoch', 100)
-
-    def _set_sparsity_level(self):
+        self.target_sparsity = params.get('sparsity_target', 0.5)
+        self.target_epoch = params.get('sparsity_target_epoch', 90)
+        self.freeze_epoch = params.get('sparsity_freeze_epoch', 100)
```
Comment (andrey-churkin) [inline nncf/common/sparsity/schedulers.py:59]: @vshampor @alexsu52 What are these default values for? Can we make these parameters required?

File: nncf/common/pruning/schedulers.py (hunk 4 above)
Comment (andrey-churkin) [reply nncf/common/pruning/schedulers.py:60]: I can rename
`target_pruning` --> `pruning_target`
`initial_pruning` --> `pruning_initial` (for consistency)
and do the sample for sparsity scheduler if needed.

But, I think, that some config parameters don't have a very good name. For example, `num_init_steps` and `pruning_steps`
(`num_warmup_epochs` and `num_pruning_epochs` are better for me).

File: nncf/common/pruning/schedulers.py (hunk 1 above)
Comment (andrey-churkin) [reply nncf/common/pruning/schedulers.py]: Done

File: nncf/common/pruning/schedulers.py (hunk 2 above)
Comment (andrey-churkin) [reply nncf/common/pruning/schedulers.py:76]: Done

File: nncf/common/pruning/schedulers.py (hunk 3 above)
Comment (andrey-churkin) [reply nncf/common/pruning/schedulers.py:178]: Done

File: nncf/common/pruning/schedulers.py (hunk 4 above)
Comment (vshampor) [reply nncf/common/pruning/schedulers.py:59]: I think you can do the corresponding renamings for `num_init_steps` and `pruning_steps` as well.

Comment (vshampor) [reply nncf/common/pruning/schedulers.py:59]: AR @andrey-churkin to rename the entities in configs in a separate PR

**Review by alexsu52 — APPROVED** [review]

File: nncf/common/sparsity/schedulers.py (hunk 7 above)
Comment (alexsu52) [reply nncf/common/sparsity/schedulers.py:58]: My opinion, `sparsity_target` and `sparsity_target_epoch` is required parameters and `sparsity_freeze_epoch` is optional parameter.

Comment (andrey-churkin) [reply nncf/common/sparsity/schedulers.py]: Let's discuss it in another PR

File: nncf/common/pruning/schedulers.py (hunk 4 above)
Comment (andrey-churkin) [reply nncf/common/pruning/schedulers.py]: After discussions, we decided to use `initial_level`, `target_level`, and `current_level` as attribute names for schedulers.

File: tests/test_algo_common.py [hunk 8]
```diff
@@ -143,7 +143,7 @@ def __str__(self):
     CompressionLevelTestStruct(
         config_provider=TestConfigCreator().add_algo('filter_pruning', {
             'num_init_steps': 1,
-            'pruning_steps': 1,
+            'pruning_steps': 2,
```
Comment (andrey-churkin) [inline tests/test_algo_common.py:146]: This test was updated because I changed the pruning schedule. From documentation
> `pruning_steps` - the number of epochs during which the pruning rate target is increased from `pruning_init` to `pruning_target` value.

i.e. on epoch with index `pruning_steps - 1` we should set `pruning_target` value.
It was not true before my PR. I fixed it.

File: tests/test_algo_common.py [hunk 9]
```diff
@@ -113,7 +113,7 @@ def __str__(self):
 magnitude_sparsity_params = {'schedule': 'multistep',
                              'multistep_steps': [1, 2],
                              'multistep_sparsity_levels': [0, 0.3, 0.5]}
-filter_pruning_params = {'schedule': 'exponential', 'num_init_steps': 0, 'pruning_steps': 2}
+filter_pruning_params = {'schedule': 'exponential', 'num_init_steps': 0, 'pruning_steps': 3}
```
Comment (andrey-churkin) [inline tests/test_algo_common.py:116]: This test was updated because I changed the pruning schedule. From documentation
> `pruning_steps` - the number of epochs during which the pruning rate target is increased from `pruning_init` to `pruning_target` value.

i.e. on epoch with index `pruning_steps - 1` we should set `pruning_target` value.
It was not true before my PR. I fixed it.

**Comment by andrey-churkin** [comment]
Jenkins please retry a build

**Review by vshampor — APPROVED** [review]
